File: legacy/typescript/engram-ts/src/subscriptions.ts
/**
 * Subscription and Notification Model for Cross-Agent Intelligence
 */

import Database from 'better-sqlite3';

/**
 * A notification about a new memory that exceeded a subscription threshold
 */
export interface Notification {
  /** Memory ID */
  memoryId: string;
  /** Namespace the memory was stored in */
  namespace: string;
  /** Memory content (for convenience) */
  content: string;
  /** Memory importance */
  importance: number;
  /** When the memory was created */
  createdAt: number;
  /** The subscription that triggered this notification */
  subscriptionNamespace: string;
  /** The threshold that was exceeded */
  threshold: number;
}

/**
 * A subscription entry
 */
export interface Subscription {
  /** Agent ID of the subscriber */
  subscriberId: string;
  /** Namespace to watch ("*" = all namespaces) */
  namespace: string;
  /** Minimum importance to trigger notification */
  minImportance: number;
  /** When this subscription was created */
  createdAt: number;
}

/**
 * Manages subscriptions and notifications
 */
export class SubscriptionManager {
  constructor(private db: Database.Database) {
    this.initTables();
  }

  private initTables(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS subscriptions (
        subscriber_id TEXT NOT NULL,
        namespace TEXT NOT NULL,
        min_importance REAL NOT NULL,
        created_at INTEGER NOT NULL,
        PRIMARY KEY (subscriber_id, namespace)
      );
      
      CREATE TABLE IF NOT EXISTS notification_cursor (
        agent_id TEXT PRIMARY KEY,
        last_checked INTEGER NOT NULL
      );
      
      CREATE INDEX IF NOT EXISTS idx_subscriptions_ns ON subscriptions(namespace);
    `);
  }

  /**
   * Subscribe an agent to a namespace
   */
  subscribe(agentId: string, namespace: string, minImportance: number): void {
    const clamped = Math.max(0, Math.min(1, minImportance));
    const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO subscriptions (subscriber_id, namespace, min_importance, created_at)
      VALUES (?, ?, ?, ?)
    `);
    
    stmt.run(agentId, namespace, clamped, Date.now() / 1000);
  }

  /**
   * Unsubscribe an agent from a namespace
   */
  unsubscribe(agentId: string, namespace: string): boolean {
    const stmt = this.db.prepare(
      'DELETE FROM subscriptions WHERE subscriber_id = ? AND namespace = ?'
    );
    const result = stmt.run(agentId, namespace);
    return result.changes > 0;
  }

  /**
   * List all subscriptions for an agent
   */
  listSubscriptions(agentId: string): Subscription[] {
    const stmt = this.db.prepare(
      'SELECT subscriber_id, namespace, min_importance, created_at FROM subscriptions WHERE subscriber_id = ?'
    );
    
    const rows = stmt.all(agentId) as Array<{
      subscriber_id: string;
      namespace: string;
      min_importance: number;
      created_at: number;
    }>;
    
    return rows.map(row => ({
      subscriberId: row.subscriber_id,
      namespace: row.namespace,
      minImportance: row.min_importance,
      createdAt: row.created_at,
    }));
  }

  /**
   * Query notifications for a subscription
   */
  private queryNotificationsForSub(
    sub: Subscription,
    since?: number,
  ): Notification[] {
    const notifications: Notification[] = [];

    if (sub.namespace === '*') {
      // All namespaces
      const stmt = since
        ? this.db.prepare(
            'SELECT id, namespace, content, importance, created_at FROM memories WHERE created_at > ? AND importance >= ?'
          )
        : this.db.prepare(
            'SELECT id, namespace, content, importance, created_at FROM memories WHERE importance >= ?'
          );

      const params = since ? [since, sub.minImportance] : [sub.minImportance];
      const rows = stmt.all(...params) as Array<{
        id: string;
        namespace: string;
        content: string;
        importance: number;
        created_at: number;
      }>;

      for (const row of rows) {
        notifications.push({
          memoryId: row.id,
          namespace: row.namespace,
          content: row.content,
          importance: row.importance,
          createdAt: row.created_at,
          subscriptionNamespace: sub.namespace,
          threshold: sub.minImportance,
        });
      }
    } else {
      // Specific namespace
      const stmt = since
        ? this.db.prepare(
            'SELECT id, namespace, content, importance, created_at FROM memories WHERE created_at > ? AND importance >= ? AND namespace = ?'
          )
        : this.db.prepare(
            'SELECT id, namespace, content, importance, created_at FROM memories WHERE importance >= ? AND namespace = ?'
          );

      const params = since
        ? [since, sub.minImportance, sub.namespace]
        : [sub.minImportance, sub.namespace];
      const rows = stmt.all(...params) as Array<{
        id: string;
        namespace: string;
        content: string;
        importance: number;
        created_at: number;
      }>;

      for (const row of rows) {
        notifications.push({
          memoryId: row.id,
          namespace: row.namespace,
          content: row.content,
          importance: row.importance,
          createdAt: row.created_at,
          subscriptionNamespace: sub.namespace,
          threshold: sub.minImportance,
        });
      }
    }

    return notifications;
  }

  /**
   * Check for notifications since last check
   * Updates the cursor so the same notifications aren't returned twice
   */
  checkNotifications(agentId: string): Notification[] {
    // Get last checked timestamp
    const cursorStmt = this.db.prepare(
      'SELECT last_checked FROM notification_cursor WHERE agent_id = ?'
    );
    const cursorRow = cursorStmt.get(agentId) as { last_checked: number } | undefined;
    const lastChecked = cursorRow?.last_checked;

    // Get agent's subscriptions
    const subscriptions = this.listSubscriptions(agentId);

    if (subscriptions.length === 0) {
      return [];
    }

    const notifications: Notification[] = [];
    const now = Date.now() / 1000;

    for (const sub of subscriptions) {
      const subNotifs = this.queryNotificationsForSub(sub, lastChecked);
      notifications.push(...subNotifs);
    }

    // Update cursor
    const updateStmt = this.db.prepare(
      'INSERT OR REPLACE INTO notification_cursor (agent_id, last_checked) VALUES (?, ?)'
    );
    updateStmt.run(agentId, now);

    // Deduplicate by memory_id
    const seen = new Set<string>();
    const unique = notifications.filter(n => {
      if (seen.has(n.memoryId)) return false;
      seen.add(n.memoryId);
      return true;
    });

    // Sort by created_at descending
    unique.sort((a, b) => b.createdAt - a.createdAt);

    return unique;
  }

  /**
   * Peek at notifications without updating cursor
   */
  peekNotifications(agentId: string): Notification[] {
    const cursorStmt = this.db.prepare(
      'SELECT last_checked FROM notification_cursor WHERE agent_id = ?'
    );
    const cursorRow = cursorStmt.get(agentId) as { last_checked: number } | undefined;
    const lastChecked = cursorRow?.last_checked;

    const subscriptions = this.listSubscriptions(agentId);

    if (subscriptions.length === 0) {
      return [];
    }

    const notifications: Notification[] = [];

    for (const sub of subscriptions) {
      const subNotifs = this.queryNotificationsForSub(sub, lastChecked);
      notifications.push(...subNotifs);
    }

    // Deduplicate
    const seen = new Set<string>();
    const unique = notifications.filter(n => {
      if (seen.has(n.memoryId)) return false;
      seen.add(n.memoryId);
      return true;
    });

    unique.sort((a, b) => b.createdAt - a.createdAt);

    return unique;
  }

  /**
   * Reset notification cursor (useful for testing or re-checking everything)
   */
  resetCursor(agentId: string): void {
    const stmt = this.db.prepare('DELETE FROM notification_cursor WHERE agent_id = ?');
    stmt.run(agentId);
  }
}

File: legacy/python/engram/subscriptions.py
"""
Subscription and Notification Model for Cross-Agent Intelligence.

Agents can subscribe to namespaces to receive notifications when new
high-importance memories are stored. This enables the CEO pattern where
a supervisor agent monitors all specialist agents without polling.

Example: CEO subscribes to all namespaces with min_importance=0.8
→ Gets notified of high-importance events from any service agent.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Optional
import sqlite3


@dataclass
class Notification:
    """A notification about a new memory that exceeded a subscription threshold."""
    memory_id: str
    namespace: str
    content: str
    importance: float
    created_at: datetime
    subscription_namespace: str
    threshold: float


@dataclass
class Subscription:
    """A subscription entry."""
    subscriber_id: str
    namespace: str          # "*" = all namespaces
    min_importance: float
    created_at: datetime


class SubscriptionManager:
    """Manages subscriptions and notifications."""

    def __init__(self, conn: sqlite3.Connection):
        """
        Create a new SubscriptionManager.

        Args:
            conn: SQLite database connection (shared with Memory)
        """
        self.conn = conn
        self._ensure_tables()

    def _ensure_tables(self):
        """Initialize subscription tables."""
        self.conn.executescript("""
            CREATE TABLE IF NOT EXISTS subscriptions (
                subscriber_id TEXT NOT NULL,
                namespace TEXT NOT NULL,
                min_importance REAL NOT NULL,
                created_at REAL NOT NULL,
                PRIMARY KEY (subscriber_id, namespace)
            );
            
            CREATE TABLE IF NOT EXISTS notification_cursor (
                agent_id TEXT PRIMARY KEY,
                last_checked REAL NOT NULL
            );
            
            CREATE INDEX IF NOT EXISTS idx_subscriptions_ns ON subscriptions(namespace);
        """)
        self.conn.commit()

    def subscribe(
        self,
        agent_id: str,
        namespace: str,
        min_importance: float,
    ) -> None:
        """
        Subscribe an agent to a namespace.

        Args:
            agent_id: The subscribing agent's ID
            namespace: Namespace to watch ("*" for all)
            min_importance: Minimum importance threshold (0.0-1.0)
        """
        clamped = max(0.0, min(1.0, min_importance))

        self.conn.execute(
            """
            INSERT OR REPLACE INTO subscriptions (subscriber_id, namespace, min_importance, created_at)
            VALUES (?, ?, ?, ?)
            """,
            (agent_id, namespace, clamped, datetime.now().timestamp())
        )
        self.conn.commit()

    def unsubscribe(self, agent_id: str, namespace: str) -> bool:
        """
        Unsubscribe an agent from a namespace.

        Args:
            agent_id: The agent ID to unsubscribe
            namespace: Namespace to unsubscribe from

        Returns:
            True if subscription was removed, False if no subscription existed
        """
        cursor = self.conn.execute(
            "DELETE FROM subscriptions WHERE subscriber_id = ? AND namespace = ?",
            (agent_id, namespace)
        )
        self.conn.commit()
        return cursor.rowcount > 0

    def list_subscriptions(self, agent_id: str) -> list[Subscription]:
        """
        List all subscriptions for an agent.

        Args:
            agent_id: The agent ID to list subscriptions for

        Returns:
            List of subscriptions
        """
        cursor = self.conn.execute(
            """
            SELECT subscriber_id, namespace, min_importance, created_at
            FROM subscriptions WHERE subscriber_id = ?
            """,
            (agent_id,)
        )

        return [
            Subscription(
                subscriber_id=row[0],
                namespace=row[1],
                min_importance=row[2],
                created_at=datetime.fromtimestamp(row[3])
            )
            for row in cursor.fetchall()
        ]

    def _query_notifications_for_sub(
        self,
        sub: Subscription,
        since: Optional[datetime] = None,
    ) -> list[Notification]:
        """Helper to query notifications for a subscription."""
        notifications = []

        if sub.namespace == "*":
            # All namespaces
            if since:
                cursor = self.conn.execute(
                    """
                    SELECT id, namespace, content, importance, created_at
                    FROM memories
                    WHERE created_at > ? AND importance >= ?
                    """,
                    (since.timestamp(), sub.min_importance)
                )
            else:
                cursor = self.conn.execute(
                    """
                    SELECT id, namespace, content, importance, created_at
                    FROM memories
                    WHERE importance >= ?
                    """,
                    (sub.min_importance,)
                )
        else:
            # Specific namespace
            if since:
                cursor = self.conn.execute(
                    """
                    SELECT id, namespace, content, importance, created_at
                    FROM memories
                    WHERE created_at > ? AND importance >= ? AND namespace = ?
                    """,
                    (since.timestamp(), sub.min_importance, sub.namespace)
                )
            else:
                cursor = self.conn.execute(
                    """
                    SELECT id, namespace, content, importance, created_at
                    FROM memories
                    WHERE importance >= ? AND namespace = ?
                    """,
                    (sub.min_importance, sub.namespace)
                )

        for row in cursor.fetchall():
            notifications.append(
                Notification(
                    memory_id=row[0],
                    namespace=row[1],
                    content=row[2],
                    importance=row[3],
                    created_at=datetime.fromtimestamp(row[4]),
                    subscription_namespace=sub.namespace,
                    threshold=sub.min_importance,
                )
            )

        return notifications

    def check_notifications(self, agent_id: str) -> list[Notification]:
        """
        Check for notifications since last check.

        Returns new memories that exceed the subscription thresholds.
        Updates the cursor so the same notifications aren't returned twice.

        Args:
            agent_id: The agent ID to check notifications for

        Returns:
            List of new notifications
        """
        # Get last checked timestamp
        cursor = self.conn.execute(
            "SELECT last_checked FROM notification_cursor WHERE agent_id = ?",
            (agent_id,)
        )
        row = cursor.fetchone()
        last_checked = datetime.fromtimestamp(row[0]) if row else None

        # Get agent's subscriptions
        subscriptions = self.list_subscriptions(agent_id)

        if not subscriptions:
            return []

        notifications = []
        now = datetime.now()

        for sub in subscriptions:
            sub_notifs = self._query_notifications_for_sub(sub, last_checked)
            notifications.extend(sub_notifs)

        # Update cursor
        self.conn.execute(
            "INSERT OR REPLACE INTO notification_cursor (agent_id, last_checked) VALUES (?, ?)",
            (agent_id, now.timestamp())
        )
        self.conn.commit()

        # Deduplicate by memory_id (in case multiple subscriptions match same memory)
        seen = set()
        unique_notifs = []
        for notif in notifications:
            if notif.memory_id not in seen:
                seen.add(notif.memory_id)
                unique_notifs.append(notif)

        # Sort by created_at descending
        unique_notifs.sort(key=lambda n: n.created_at, reverse=True)

        return unique_notifs

    def peek_notifications(self, agent_id: str) -> list[Notification]:
        """
        Peek at notifications without updating cursor.

        Args:
            agent_id: The agent ID to peek notifications for

        Returns:
            List of notifications (cursor is NOT updated)
        """
        # Get last checked timestamp
        cursor = self.conn.execute(
            "SELECT last_checked FROM notification_cursor WHERE agent_id = ?",
            (agent_id,)
        )
        row = cursor.fetchone()
        last_checked = datetime.fromtimestamp(row[0]) if row else None

        subscriptions = self.list_subscriptions(agent_id)

        if not subscriptions:
            return []

        notifications = []

        for sub in subscriptions:
            sub_notifs = self._query_notifications_for_sub(sub, last_checked)
            notifications.extend(sub_notifs)

        # Deduplicate
        seen = set()
        unique_notifs = []
        for notif in notifications:
            if notif.memory_id not in seen:
                seen.add(notif.memory_id)
                unique_notifs.append(notif)

        # Sort by created_at descending
        unique_notifs.sort(key=lambda n: n.created_at, reverse=True)

        return unique_notifs

    def reset_cursor(self, agent_id: str) -> None:
        """
        Reset notification cursor (useful for testing or re-checking everything).

        Args:
            agent_id: The agent ID to reset cursor for
        """
        self.conn.execute(
            "DELETE FROM notification_cursor WHERE agent_id = ?",
            (agent_id,)
        )
        self.conn.commit()
